flake: add Nixpkgs default arguments as input

roberth opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Actually solve the problem of `nixpkgs.config` in a flake, and more.

Flakes are just functional programming with default arguments. The fact that everything has a default might throw you off, but it's still functional programming, where `follows` is apply. Enjoy!

Here's a flake producing `skype` from `legacyPackages` without the unfree error:

`flake.nix`
```nix
{
  description = "A very basic flake";

  inputs.nixpkgs.url = "github:hercules-ci/nixpkgs/add-flake-default-arguments-as-input";
  inputs.nixpkgs.inputs.parameters.follows = "parameters";
  inputs.parameters.url = "./parameters";

  outputs = { self, nixpkgs, ... }: {

    packages.x86_64-linux.skype = nixpkgs.legacyPackages.x86_64-linux.skype;

    defaultPackage.x86_64-linux = self.packages.x86_64-linux.skype;

  };
}
```

`parameters/flake.nix`
```nix
{
  outputs = { self }: {
    lib.pkgsParameters = {
      overlays = [ ];
      buildSystem = null; # same as host system
      config = {
        allowUnfree = true;
      };
    };
  };
}
```

Refs

- https://github.com/NixOS/nixpkgs/issues/158740
- https://github.com/NixOS/nixpkgs/issues/109421
- https://github.com/NixOS/nixpkgs/pull/117842
- https://github.com/NixOS/nixpkgs/issues/101101
- https://github.com/NixOS/nixpkgs/pull/101367

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

roberth committed:
flake: add Nixpkgs default arguments as input

otavio reviewed (commented):
Oh! That's nice :-D

Ma27 reviewed (commented):
Some docs in e.g. the nixpkgs manual may be nice, but other than that, I like the idea %)

roberth committed:
doc/using/configuration: Add "Configure with Flakes"

roberth committed:
doc/using/configuration: Move allowUnfree examples into own section

This way the disclaimer is more effective and the preceding docs
are more focused.

roberth committed:
doc/using/configuration: Add headers for config methods

pasqui23 commented on flake.nix:
```diff
@@ -3,7 +3,9 @@
   description = "A collection of packages for the Nix package manager";
 
-  outputs = { self }:
+  inputs.parameters.url = "github:NixOS/nixpkgs/default-parameters";
```
```suggestion
  inputs.parameters.url = ./default-params;
```

viperML commented:
I am eagerly waiting for this (or any other similar solution) to be merged.

As mentioned in the PR, it would solve a common problem of flakes, that is to provide packages with unfree licenses. The current status is either to:
1. Provide package outputs that cannot be built (bad)
2. Instantiate our own nixpkgs, without any prompt of the user (bad)

A flake with unfree outputs could just bundle its default nixpkgs config, that allows for `allowUnfree = true`, while also being able to follow the user's config, eliminating problem 2.

--- 
Another usecase that comes to my mind is using `config.contentAddressedByDefault = true`. As a flake writer, I need to use `nixpkgs.legacyPackages` so my consumers can bring their own nixpkgs with their config. Even then, they would have to use an intermediate flake to wrap nixpkgs, such that it would import nixpkgs with its config, then re-export it, and then set the follows.

Consuming flake -> nixpkgs with contentAddressedByDefault -> My flake

In a similar fashion to https://github.com/numtide/nixpkgs-unfree , but for every combination of configuration options...

edolstra commented:
Strong :-1: on this. We should not invent yet another undiscoverable ad-hoc configuration mechanism to work around the lack of a configuration mechanism in flakes. Flake inputs are not intended to pass configuration data around and we should not start using them for that.

We should address the flake configuration issue at some point (probably after stabilising current flakes), but ideally in a standardized, discoverable way.

edolstra commented on flake.nix:
```diff
@@ -3,7 +3,9 @@
   description = "A collection of packages for the Nix package manager";
 
-  outputs = { self }:
+  inputs.parameters.url = "github:NixOS/nixpkgs/default-parameters";
```
This refers to the "default-params" branch of the nixpkgs repo, which doesn't exist AFAIK and whose `flake.nix` presumably is the top-level flake. Did you mean `github:NixOS/nixpkgs?dir=default-params` or something like that?

roberth replied:
I thought I created the branch. At the time, a branch seemed more suitable, because before the source tree abstraction branch, a relative path subflake still required a nar hash that matches the local file contents, whereas an input from a branch was more easily understood and more stable in terms of its implementation in the nix tool.

The default parameters were "empty" values like `[ ]` for overlays and `{ }` for `config`, and the default behavior emerges from having empty "overrides". So the defaults for these values wouldn't ever have to change, and the branch would most likely have only one commit referencing a tiny tree that's unrelated to trees in the normal branches. Putting it in a separate repo may have been more appropriate.

roberth commented:
> We should not invent yet another undiscoverable ad-hoc configuration mechanism

I agree with this part. One of the valuable lessons of NixOS is that a simple language intended for package management is powerful enough to support all configuration needs. By restricting the configuration system to a small but powerful set of primitives, being the Nix language, we have gained a system with a small core that provides a consistent paradigm for configuration that allows the interoperation of both highly capable modules (NixOS) and easy to use configurations (configuration.nix).

> the lack of a configuration mechanism in flakes

So one thing that NixOS teaches us is that function parameters, and by extension the attributes in them (`config.*`, etc) are suitable for configuration. We didn't need "configurable expressions" and we were better off without them, as their lack simplifies the language.

What are flake inputs if not parameters with defaults? The only thing about inputs that's really new is that their defaults come from a lock file, and this is great. As for the other aspects, they are a reinvention of concepts that already existed in the language, such as function application, but without interaction with critical other language features such as scoping.

What has happened is that in the name of standardization, we have let go of our ideal of a small language, and now we're starting to see the paradigm falling apart, in the form of a claim that "inputs" are not suitable for "configuration", whereas previously these were the same thing, interoperating in harmony with the rest of the language.

**A warning from the Java world**

To understand what Flakes are, we can decompose them into
 - a system for retrieving and locking dependencies
 - a set of CLI improvements
 - a dependency injection system

Dependency injection systems are the bread and butter of the Java world. These are systems that solve the "problem" of invoking the constructors of singleton-like objects and passing configuration parameters into them - ie exactly what flakes are doing. They have many frameworks to fulfill this task, but it is noteworthy that the Spring framework has gone full circle and reverted back to using _regular code_ for this purpose, albeit annotated code.

Both the sheer number of implementations and the reversal to regular code are indications that dependency injection is not solvable without the occasional need for a programming language to specify what goes where.

If the JVM enforced a single dependency injection framework (like Nix does with the hardcoded `call-flake.nix`), neither the experimentation with frameworks, nor the reversal to "inject it yourself" would have been possible.

It's not too late to simplify flakes. It's still experimental.

I regret not taking part in the discussions and early design in 2018. I did not foresee that Flakes would be as invasive or that its design wouldn't be improved much from what was presented at that NixCon. My hindsight is 20/20.

roberth commented:
> It was not intended as a dependency injection system.

Then it should not invoke the `outputs` function, and certainly not for the outputs of any of the inputs.

> I don't think there is that much that can be simplified further [...] The only major thing to reconsider is the input override mechanism

If you don't invoke the `outputs`, the user can assemble their inputs' outputs in any way they like using the Nix language, making the input override and follows mechanisms redundant. Follows has been a source of bugs, and I'm still getting weird behavior when I update an input flakeref sometimes. Good riddance?

> since the flake spec is already pretty minimal.

Where do I find the latest spec?

> functions (at least as done in Nix) are not suitable for configuration [...] NixOS module system

Yet the module system fits within the functional paradigm, which illustrates the point that a small Nix is good and supports both ease of use (config modules) and fairly complicated transformations (functions in modules). The same can apply to a more minimal implementation of the flake goals; properly excluding dependency injection.

SomeoneSerge commented:
Maybe a compromise would be to create a proxy flake repo under `nix-community` akin to `nixpkgs-unfree`, which would import `nixpkgs` with an overridable `config`. On one hand, this would read as less of an endorsement, on the other it'd further discourage users from treating flake inputs as paths (incl. `callPackage "${nixpkgs}/pkgs/...`).

I think @roberth's proposal is very important. The "flakes as paths" pattern is already being commonly abused. E.g., at the time of writing, `github:ggerganov/llama.cpp` imports Nixpkgs almost a dozen times with different configs to expose attributes for different CUDA and ROCm targets, because there isn't a flake-friendly alternative.

I also think it's a fair remark that since it isn't exactly trivial to treat `flake.nix` as a normal Nix function, flakes "as used in practice" are more than a way to pin dependencies. Flakes may not be optimal as a DI system, and that may be a reason not to promote them as such. However, if we're choosing between configs as flake inputs and flake inputs as paths, and there's still no prospect of a third option, I think we should choose the former.

@edolstra I wonder how you feel on the matter now that some time has passed?

---

Regarding the implementation and mirroring some comments from the nixos matrix: since flakes weren't designed to be a DI system, they aren't particularly strong in this role. E.g. if one wanted to use `nixpkgs` and `flake-parts`, and tried additionally to minimize the flake graph, they'd likely have a trouble making `flake-parts`' [`nixpkgs-lib`](https://github.com/hercules-ci/flake-parts/blob/07f6395285469419cf9d078f59b5b49993198c00/flake.nix#L5) use the same revision of `nixpkgs` as the parent flake.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/im-not-convinced-flake-inputs-should-exist/50636/16

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-define-the-default-package-for-a-nixos-module-in-a-nix-flake/77311/12
